# Paquetes

In [ ]:
install.packages("readxl")
library(readxl)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
install.packages("glmnet")
library(glmnet)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Warning message in install.packages("glmnet"):
“installation of package ‘glmnet’ had non-zero exit status”


In [ ]:
install.packages("boot")
library(boot)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
install.packages("stats")
library(stats)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Warning message:
“package ‘stats’ is a base package, and should not be updated”


In [ ]:
install.packages("caret")
library(caret)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
install.packages("gplots")
library(gplots)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



#(a) Carga de datos

In [ ]:
library(readxl)
df <- read_excel("body.xls")
colnames(df)<- (c("BIAC","BIIL","BITRO","CHEST1","CHEST2","ELBOW","WRIST", "KNEE","ANKLE","SHOUL","CHESTG","WAISTG","NAVEL","HIP","GLUTE","BICEP", "FLOREA","KNEEG","CALF","ANKLEG","WRISTG","AGE","WEIG","HEIG","GEN"))

New names:
• `26` -> `26...2`
• `26` -> `26...17`


In [ ]:
df

BIAC BIIL BITRO CHEST1 CHEST2 ELBOW WRIST KNEE ANKLE SHOUL ⋯ BICEP FLOREA
1   43.7 28.5 33.5  16.9   30.8   14.0  11.8  20.6 15.1  110.5 ⋯ 34.4  28.0  
2   40.1 28.2 33.3  20.9   31.7   13.9  10.9  19.7 14.1  115.1 ⋯ 33.4  28.8  
3   44.3 29.9 34.0  18.4   28.2   13.9  11.2  20.9 15.0  104.5 ⋯ 31.0  26.2  
4   42.5 29.9 34.0  21.5   29.4   15.2  11.6  20.7 14.9  107.5 ⋯ 32.0  28.4  
5   43.3 27.0 31.5  19.6   31.3   14.0  11.5  18.8 13.9  119.8 ⋯ 33.0  28.0  
6   43.5 30.0 34.0  21.9   31.7   16.1  12.5  20.8 15.6  123.5 ⋯ 42.4  32.3  
7   44.4 29.8 33.2  21.8   28.8   15.1  11.9  21.0 14.6  120.4 ⋯ 34.1  28.0  
8   43.5 26.5 32.1  15.5   27.5   14.1  11.2  18.9 13.2  111.0 ⋯ 33.0  26.0  
9   42.0 28.0 34.0  22.5   28.0   15.6  12.0  21.1 15.0  119.5 ⋯ 36.5  29.2  
10  40.3 29.0 33.0  20.1   30.3   13.4  10.4  19.4 14.5  117.1 ⋯ 34.6  27.9  
11  43.7 29.0 31.3  20.5   29.7   15.0  11.7  20.9 16.0  123.5 ⋯ 38.5  30.4  
12  47.4 29.6 35.7  20.8   31.4   16.1  11.3  21.5 15.4  116.5 ⋯ 33.5  29.0  
13  40.3 27.5 31.4  21.7   28.0   13.3  10.3  18.8 13.2  113.0 ⋯ 35.6  29.0  
14  41.0 26.8 32.2  21.9   28.6   14.9  10.6  17.8 14.0  107.5 ⋯ 36.0  29.0  
15  45.0 27.0 33.2  21.7   30.6   13.7  11.1  20.7 14.0  112.0 ⋯ 34.5  29.5  
16  39.9 30.0 34.5  21.0   29.4   15.6  11.9  21.2 16.0  112.2 ⋯ 37.0  28.8  
17  43.0 26.5 30.3  19.3   30.0   14.8  11.2  19.7 14.7  120.0 ⋯ 31.2  26.9  
18  43.1 28.6 33.4  22.2   29.5   14.9  12.2  20.8 14.8  109.0 ⋯ 34.5  28.5  
19  43.6 29.3 34.4  20.2   32.6   15.4  10.9  20.7 15.5  118.5 ⋯ 37.0  30.5  
20  42.0 27.5 30.7  21.3   32.0   13.1  11.1  19.2 13.9  116.0 ⋯ 32.0  27.6  
21  43.8 28.0 33.3  20.0   32.0   15.0  11.5  20.4 14.4  111.0 ⋯ 37.0  30.0  
22  42.3 26.4 31.2  18.0   30.9   14.6  10.8  18.6 13.8  117.7 ⋯ 31.2  26.9  
23  42.7 29.9 35.0  21.8   32.8   14.3  11.2  19.8 14.1  123.9 ⋯ 37.0  28.3  
24  44.8 27.8 32.2  18.3   31.5   15.2  11.6  19.4 14.7  120.6 ⋯ 39.4  31.9  
25  46.0 30.1 34.5  20.2   31.1   16.4  13.3  22.2 14.9  129.5 ⋯ 37.5  30.1  
26  45.4 31.8 35.2  20.2   32.3   14.6  10.5  20.2 15.3  115.0 ⋯ 35.5  28.5  
27  40.5 28.3 33.4  19.2   28.8   14.6  11.1  20.8 14.5  116.0 ⋯ 32.5  27.8  
28  39.4 25.5 30.2  17.6   27.7   13.0  10.2  18.9 13.2  107.8 ⋯ 31.2  26.5  
29  40.2 27.2 31.7  18.1   26.5   13.3  10.1  18.6 13.2  100.2 ⋯ 27.5  24.8  
30  44.2 30.3 34.7  19.4   30.0   14.9  11.0  19.1 15.8  113.0 ⋯ 32.8  26.2  
⋮   ⋮    ⋮    ⋮     ⋮      ⋮      ⋮     ⋮     ⋮    ⋮     ⋮     ⋱ ⋮     ⋮     
477 35.6 27.3 31.7  17.5   27.3   12.4   9.8  17.6 12.7  102.2 ⋯ 31.1  24.4  
478 37.8 29.0 32.4  20.6   28.5   12.7   9.1  18.2 13.2  110.3 ⋯ 29.9  24.9  
479 40.9 28.7 29.2  18.2   27.3   12.9   9.6  16.8 13.4  107.3 ⋯ 28.8  23.2  
480 34.0 25.9 29.4  20.2   24.2   11.4   9.6  16.6 12.6   95.0 ⋯ 29.3  23.1  
481 34.9 31.3 31.7  20.9   31.5   13.0   9.6  17.8 13.7  115.6 ⋯ 35.7  25.6  
482 35.8 31.5 32.6  17.5   25.2   12.9   9.8  17.9 13.4   99.3 ⋯ 27.0  22.3  
483 36.5 27.5 30.8  18.5   24.5   11.6   9.6  17.0 11.7   93.1 ⋯ 24.4  22.1  
484 32.9 26.8 28.7  16.3   24.4   11.0   8.6  16.4 12.6   94.9 ⋯ 24.9  22.3  
485 36.9 28.5 31.5  15.6   27.8   11.6   9.8  17.2 12.9  100.0 ⋯ 25.6  23.0  
486 35.8 29.4 32.0  18.0   25.9   11.8   9.8  18.0 12.7   98.6 ⋯ 27.3  24.0  
487 39.4 28.3 30.2  21.6   29.7   12.8  10.5  19.5 14.4  107.2 ⋯ 29.9  25.0  
488 36.5 28.3 29.7  17.7   26.4   12.0  10.0  17.1 12.8   98.3 ⋯ 27.6  23.3  
489 35.8 27.3 28.5  18.2   23.5   11.2  10.2  17.5 12.9   92.4 ⋯ 25.5  23.0  
490 34.4 29.7 31.7  17.5   24.7   12.6  10.2  17.7 12.4   93.0 ⋯ 31.0  26.3  
491 36.5 26.6 30.8  16.0   24.9   12.6  10.7  17.0 13.0   99.5 ⋯ 28.3  24.7  
492 38.5 25.6 31.7  17.0   25.6   11.8  10.2  16.8 12.9   92.8 ⋯ 28.1  23.1  
493 35.3 25.4 28.5  15.1   24.9   10.6  10.4  17.0 12.6   95.9 ⋯ 26.8  22.8  
494 36.2 28.7 27.8  16.0   23.7   12.4  10.2  17.9 13.2   96.1 ⋯ 26.2  22.6  
495 34.9 28.7 32.9  16.8   26.8   13.4  10.8  17.9 12.9   98.5 ⋯ 29.6  25.1  
496 38.7 29.7 32.0  15.8  

In [ ]:
#recorte en 2 dataframes
fem  = df[df$GEN==0, ]
masc = df[df$GEN==1, ]


#(b) Bootstrap (intervalo de confianza de la mediana de WEIG)

In [ ]:
#Bootstrap Normal
estimar_se_mediana <- function(data, N = 1000){
  titahatboot <- rep(0, N)
  for(i in 1:N){
    Xboot <- sample(data, length(data), replace = TRUE)
    titahatboot[i] <- median(Xboot)
  }
  sqrt(mean((titahatboot-mean(titahatboot))^2))
}

#Intervalo bootsrap normal
bootstrap_normal <- function (data, n) {
  mediana_muestra <- median(data)
  se_boot <- estimar_se_mediana(data, n)
  intervalo_boot <- c(mediana_muestra - 1.96*se_boot, mediana_muestra + 1.96*se_boot)
  return(intervalo_boot)
}



#Intervalo Bootstrap Percentil
bootstrap_percentil <- function(data, n) {
  mediana_original <- median(data)
  bootstrap_samples <- matrix(sample(data, n * length(data), replace = TRUE), n)
  mediana_bootstrap <- apply(bootstrap_samples, 1, median)

  sesgo_bootstrap <- mediana_bootstrap - mediana_original  # Corrección del sesgo
  intervalo_confianza <- quantile(sesgo_bootstrap, c(0.025, 0.975))

  return(mediana_original + intervalo_confianza)
}

In [ ]:
cat('La mediana estimada del peso para el género masculino es', median(masc$WEIG), 'kg', '\n')
cat('La mediana estimada del peso para el género femenino es', median(fem$WEIG), 'kg')


La mediana estimada del peso para el género masculino es 77.5 kg 
La mediana estimada del peso para el género femenino es 59 kg

In [ ]:
set.seed(123)
n_bootstrap <- 1000  # repeticiones bootstrap

ic_estandar_masc<- bootstrap_estandar(masc$WEIG, n_bootstrap)
ic_estandar_fem <- bootstrap_estandar(fem$WEIG, n_bootstrap)

cat('Intervalo de confianza de nivel 0.95 (estandar) para hombres: [',ic_estandar_masc[1],',',ic_estandar_masc[2],']', '\n')
cat('Intervalo de confianza de nivel 0.95 (estandar) para mujeres: [',ic_estandar_fem[1],',',ic_estandar_fem[2],']')


ERROR: ignored

In [ ]:
set.seed(123)
n_bootstrap <- 1000  # repeticiones bootstrap

ic_normal_masc<- bootstrap_normal(masc$WEIG, n_bootstrap)
ic_normal_fem <- bootstrap_normal(fem$WEIG, n_bootstrap)

cat('Intervalo de confianza de nivel 0.95 (normal) para hombres: [',ic_normal_masc[1],',',ic_normal_masc[2],']', '\n')
cat('Intervalo de confianza de nivel 0.95 (normal) para mujeres: [',ic_normal_fem[1],',',ic_normal_fem[2],']')


In [ ]:
ic_perc_masc <- bootstrap_percentil(masc$WEIG, n_bootstrap)
ic_perc_fem  <- bootstrap_percentil(fem$WEIG, n_bootstrap)


cat('Intervalo de confianza de nivel 0.95 (método percentil) para hombres: [',ic_perc_masc[1],',',ic_perc_masc[2],']', '\n')
cat('Intervalo de confianza de nivel 0.95 (método percentil) para mujeres: [',ic_perc_fem[1],',',ic_perc_fem[2],']')


#print(ic_perc_masc)
#print(ic_perc_fem)

La estimación puntual se realiza en ambos casos a partir de la mediana muestral, por lo que resulta ser de 77.5 y 59 kg respectivamente.

Los intervalos de confianza de nivel 95% para la mediana, obtenidos por ambos métodos (asumiendo normalidad de la distribución del estimado, o bien sin asumirlo) resultan ser muy similares, ya que aprozimando a 2 cifras significativas quedan:
mediana de peso masculino: [75; 80] (método normal) y [76; 80] (método cuantiles)
mediana de peso femenino: [58; 60] (para ambos métodos)

Esto podría indicar que para el tamaño de muestra con el que trabajamos, la distribución normal parece ser una buena aproximación para la distribución del estimador.



#(c) Gráficos de dispersión de WEIG y HEIG por género

In [ ]:


#para que sean 2 gráficos, uno al lado del otro
par(mfrow = c(1, 2))
options(repr.plot.width = 15, repr.plot.height = 5)

plot(masc$HEIG, masc$WEIG,
     #col = ifelse(df$D == "Masculino", "blue", "red"), pch = 16,
     xlab = "Altura (cm)", ylab = "Peso (kg)", main = "Dispersión de altura vs peso (género MASCULINO)")

# Agregar leyenda
#legend("topright", legend = levels(df$D), col = c("blue", "red"), pch = 16, title = "Género")

plot(fem$HEIG, fem$WEIG,
     #col = ifelse(df$D == "Masculino", "blue", "red"), pch = 16,
     xlab = "Altura (cm)", ylab = "Peso (kg)", main = "Dispersión de altura vs peso (género FEMENINO)")



En ambos casos el gráfico de dispersión sugiere una correlación lineal positiva entre las variables peso y altura, es decir, a mayor peso se observa una tendencia a mayor altura. Esta relación además parece ser lineal. La dispersión de peso parece ser independiente de las alturas.

#(d) Gráficos de dispersión de WEIG y HEIG por género con línea (ksmooth)

In [ ]:
#para que sean 2 gráficos, uno al lado del otro
par(mfrow = c(1, 2))
options(repr.plot.width = 15, repr.plot.height = 5)


plot(masc$HEIG, masc$WEIG,
     #col = ifelse(df$D == "Masculino", "blue", "red"), pch = 16,
     xlab = "Altura (cm)", ylab = "Peso (kg)", main = "Dispersión de altura vs peso (género MASCULINO)")

lines(ksmooth(masc$HEIG, masc$WEIG, kernel = "normal", bandwidth = 10,
        #range.x = range(x),
        #n.points = max(100L, length(x)), x.points
        )
      )

# Ajustar un modelo de regresión lineal
modelo_lm <- lm(WEIG ~ HEIG, data = masc)

# Dibujar la recta de cuadrados mínimos
abline(modelo_lm, col = "green", lwd = 2)


# Agregar leyenda
#legend("topright", legend = levels(df$D), col = c("blue", "red"), pch = 16, title = "Género")

plot(fem$HEIG, fem$WEIG,
     #col = ifelse(df$D == "Masculino", "blue", "red"), pch = 16,
     xlab = "Altura (cm)", ylab = "Peso (kg)", main = "Dispersión de altura vs peso (género FEMENINO)")
lines(ksmooth(fem$HEIG, fem$WEIG, kernel = "normal", bandwidth = 10,
        #range.x = range(x),
        #n.points = max(100L, length(x)), x.points
        )
      )


# Ajustar un modelo de regresión lineal
modelo_lm <- lm(WEIG ~ HEIG, data = fem)

# Dibujar la recta de cuadrados mínimos
abline(modelo_lm, col = "green", lwd = 2)



La relación en ambos casos parece ser lineal, ya que la curva obtenida por ksmooth (y ventana h=10) tiene muy poca curvatura y parece prácticamente recta; más aun la del género femenino que la del masculino.

#(e) Leave one out para hallar ventana óptima

In [ ]:
loocv_bw <- function(data, x_col, y_col, bw_values) {
  n <- nrow(data)
  mse_values <- numeric(length(bw_values))

  for (i in seq_along(bw_values)) {
    mse <- 0

    for (j in 1:n) {
      # Deja fuera un punto en cada iteración para LOOCV
      data_leave_one_out <- data[-j, ]

      # Ajusta el modelo con ksmooth
      smooth_fit <- ksmooth(data_leave_one_out[[x_col]], data_leave_one_out[[y_col]],kernel = "normal", bandwidth = bw_values[i], x.points = data[j, ][[x_col]])

      # Predice el valor para el punto excluido
      predicted_value <- smooth_fit$y[1]

      y_facto = data[j, ][[y_col]]


    #  if(is.na(y_fact)){print('yfact es NA')} #debug
    #  if(is.na(predicted_value)){print(cat('predicted_value es NA','y j es:',j))}

      # Calcula el error cuadrático
      #mse <- mse + (y_fact - predicted_value)^2

      if(!is.na(y_facto) && !is.na(predicted_value)){
        # Calcula el error cuadrático
        mse <- mse + (y_facto - predicted_value)^2
      }
    }

    # Calcula el error cuadrático medio para el bandwidth
    #mse_values[i] <- mse / n
    mse_values[i] <- mse / sum(!is.na(data[, y_col]))
    #print(sum(!is.na(data[, y_col])))
  }

  # Encuentra el índice del bandwidth con el menor error cuadrático medio
  best_bw_index <- which.min(mse_values)
  best_bw <- bw_values[best_bw_index]

  return(list(best_bandwidth = best_bw, mse_values = mse_values))
}

In [ ]:
expandida <- expand.grid(bandwidth = bw_values)
#colnames(bw_values) <- c("mtry")

In [ ]:
colnames(expandida) <- ("mtry")

In [ ]:
loocv_bw_caret <- function(data, x_col, y_col, bw_values) {
  formula <- as.formula(paste0(y_col, " ~ ", x_col))

  loocv_caret <- train(
    formula,
    data = data,
    method = 'rf',
    #method = 'ksmooth',
    tuneGrid = expandida,
    trControl = trainControl(method = 'LOOCV', summaryFunction = defaultSummary) #por default usa MSE
  )

  return(loocv_caret)
}

In [ ]:
loocv_bw_caret <- function(data, x_col, y_col, bw_values) {
  formula <- as.formula(paste0(y_col, " ~ ", x_col))

  loocv_caret <- train(
    formula,
    data = data,
    method = 'lm',
    #method = 'ksmooth',
    tuneGrid = expand.grid(bandwidth = bw_values),
    trControl = trainControl(method = 'LOOCV', summaryFunction = defaultSummary) #por default usa MSE
  )

  return(loocv_caret)
}

In [ ]:
'''
loocv_bw_caret <- function(data, x_col, y_col, bw_values) {
  loocv_caret <- train(
    x = data[, x_col],
    y = data[, y_col],
    method = 'ksmooth',
    tuneGrid = expand.grid(bandwidth = bw_values),
    trControl = trainControl(method = 'LOOCV', summaryFunction = defaultSummary) #por default usa MSE
  )

  return(loocv_caret)
}
'''

In [ ]:
bw_values <- seq(5, 20, by = 0.5)



result_masc <- loocv_bw(masc, 'WEIG', 'HEIG', bw_values)
cat("Bandwidth Optimo(masc):", result_masc$best_bandwidth, "\n")
cat("Errores cuadráticos medios(masc):", result_masc$mse_values, "\n")

result_fem <- loocv_bw(fem, 'WEIG', 'HEIG', bw_values)
cat("Bandwidth Optimo(fem):", result_fem$best_bandwidth, "\n")
cat("Errores cuadráticos medios(fem):", result_fem$mse_values, "\n")

Bandwidth Optimo(masc): 12 
Errores cuadráticos medios(masc): 38.44017 38.86794 38.67893 38.5219 38.39098 38.28101 38.19046 38.11678 38.05835 38.00728 37.96954 37.94211 37.92441 37.91566 37.9156 37.92386 37.94039 37.96403 37.99509 38.03483 38.08171 38.13455 38.19249 38.25837 38.32918 38.40447 38.48573 38.57093 38.65981 38.75258 38.84887 
Bandwidth Optimo(fem): 8 
Errores cuadráticos medios(fem): 33.58439 33.4802 33.38363 33.30268 33.24523 33.21406 33.20865 33.22773 33.26868 33.32852 33.40408 33.49249 33.59147 33.69867 33.81203 33.93005 34.05159 34.17495 34.29945 34.42408 34.54839 34.67155 34.79289 34.91267 35.03067 35.14621 35.25955 35.3705 35.47931 35.58601 35.69054 


In [ ]:
install.packages("randomForest")

In [ ]:
install.packages("loess")

In [ ]:
library(randomForest)

In [ ]:
#no sé por qué no funciona

masc_limpio <- na.omit(masc[complete.cases(masc$WEIG, masc$HEIG), ])
fem_limpio <- na.omit(fem[complete.cases(fem$WEIG, fem$HEIG), ])

result_masc_caret <- loocv_bw_caret(masc_limpio, "WEIG", "HEIG", bw_values)
result_masc_caret
#cat("Bandwidth Optimo(masc):", result_masc$best_bandwidth, "\n")
#cat("Errores cuadráticos medios(masc):", result_masc$mse_values, "\n")

result_fem_caret <- loocv_bw_caret(fem_limpio, 'WEIG', 'HEIG', bw_values)
result_fem_caret
#cat("Bandwidth Optimo(fem):", result_fem$best_bandwidth, "\n")
#cat("Errores cuadráticos medios(fem):", result_fem$mse_values, "\n")

In [ ]:
#No parece ser un problema de clase ni de NA:

class(masc$HEIG)
class(masc$WEIG)
class(fem$HEIG)
class(fem$WEIG)

sum(is.na(masc$HEIG))
sum(is.na(masc$WEIG))
sum(is.na(fem$HEIG))
sum(is.na(fem$WEIG))

In [ ]:
loocv_bw_final <- function(data,bw_values){
  n <- length(data$HEIG)
  mse_values <- c()
  data_part <- data.frame(HEIG = data$HEIG, WEIG = data$WEIG)

  for (bw in bw_values){
    mse <- 0
    for (i in 1:n){
      modelo_ajustado <- ksmooth(data_part$HEIG[-i], data_part$WEIG[-i], kernel = "normal", bandwidth = bw, x.points = df_part$HEIG[i])
      mse <- mse + (data_part$WEIG[i] - modelo_ajustado$y[1])^2
    }
    mse_values <- c(mse_values , mse/n)
  }

  bw_optimo = bw_values[which.min(mse_values)]
  ecm_optimo = min(mse_values)

  return(bw_optimo)
}


#plot(windows, errores_w, main = "Busqueda de mejor ventana segun error cuadratico en hombres", xlab= "Tamaño ventana", ylab = "Error Cuadratico") +
#  points(windows[which.min(errores_w)], min(errores_w), col = "red", pch = 19)

In [ ]:
#Falta:

#e) que funcione lo de caret y ver si nos da lo mismo (otros grupos?)



#(f) Gráficos de dispersión de WEIG y HEIG por género (con ksmooth de ventana óptima)

In [ ]:
#para que sean 2 gráficos, uno al lado del otro
par(mfrow = c(1, 2))
options(repr.plot.width = 15, repr.plot.height = 5)


plot(masc$HEIG, masc$WEIG,
     #col = ifelse(df$D == "Masculino", "blue", "red"), pch = 16,
     xlab = "Altura (cm)", ylab = "Peso (kg)", main = "Dispersión de altura vs peso (género MASCULINO)")

lines(ksmooth(masc$HEIG, masc$WEIG, kernel = "normal", bandwidth = result_masc$best_bandwidth,
        #range.x = range(x),
        #n.points = max(100L, length(x)), x.points
        )
      )


# Ajustar un modelo de regresión lineal
modelo_lm <- lm(WEIG ~ HEIG, data = masc)

# Dibujar la recta de cuadrados mínimos
abline(modelo_lm, col = "green", lwd = 2)




# Agregar leyenda
#legend("topright", legend = levels(df$D), col = c("blue", "red"), pch = 16, title = "Género")

plot(fem$HEIG, fem$WEIG,
     #col = ifelse(df$D == "Masculino", "blue", "red"), pch = 16,
     xlab = "Altura (cm)", ylab = "Peso (kg)", main = "Dispersión de altura vs peso (género FEMENINO)")
lines(ksmooth(fem$HEIG, fem$WEIG, kernel = "normal", bandwidth = result_fem$best_bandwidth,
        #range.x = range(x),
        #n.points = max(100L, length(x)), x.points
        )
      )

# Ajustar un modelo de regresión lineal
modelo_lm <- lm(WEIG ~ HEIG, data = fem)

# Dibujar la recta de cuadrados mínimos
abline(modelo_lm, col = "green", lwd = 2)


In [ ]:
# FALTA agegar recta de cuadrados minimo en los graficos (con referencias de colores) + sacar conclusiones

In [ ]:
'''##funcion sin parche para NA(deberia andar si soluciono ksmooth)
loocv_bw <- function(data, x_col, y_col, bw_values) {
  n <- nrow(data)
  mse_values <- numeric(length(bw_values))

  for (i in seq_along(bw_values)) {
    mse <- 0

    for (j in 1:n) {
      # Deja fuera un punto en cada iteración para LOOCV
      data_leave_one_out <- data[-j, ]

      # Ajusta el modelo con ksmooth
      smooth_fit <- ksmooth(data_leave_one_out[[x_col]], data_leave_one_out[[y_col]], bandwidth = bw_values[i])

      # Predice el valor para el punto excluido
      predicted_value <- smooth_fit$y[j]

      # Calcula el error cuadrático
      mse <- mse + (data[j, ][[y_col]] - predicted_value)^2
    }

    # Calcula el error cuadrático medio para el bandwidth
    mse_values[i] <- mse / n
  }

  # Encuentra el índice del bandwidth con el menor error cuadrático medio
  best_bw_index <- which.min(mse_values)
  best_bw <- bw_values[best_bw_index]

  return(list(best_bandwidth = best_bw, mse_values = mse_values))
}
'''

#(g) Primer modelo lineal (usando todas las variables)

In [ ]:
#Separación en train y test

#Leemos el archivo de TrainTest
train_test <- read.table("TrainTest.txt", header = TRUE)
colnames(train_test)<-c('TrainTest')

# Agregamos columna a df
df$TrainTest <- train_test$TrainTest

# Separamos subconjuntos de datos para entrenamiento y prueba
datos_entrenamiento <- df[df$TrainTest, ]
datos_prueba <- df[!df$TrainTest, ]

In [ ]:
#no tenemos que poner GEN como un factor?? (preguntar a otros grupos?)

In [ ]:
# Seleccionamos todas las variables predictoras (y la variable de respuesta)
predictores <- c("BIAC","BIIL","BITRO","CHEST1","CHEST2","ELBOW","WRIST", "KNEE","ANKLE","SHOUL","CHESTG","WAISTG","NAVEL","HIP","GLUTE","BICEP", "FLOREA","KNEEG","CALF","ANKLEG","WRISTG","AGE","HEIG","GEN")
respuesta <- 'WEIG'

# Crea el modelo lineal
modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+")), data = datos_entrenamiento)

# Muestra un resumen del modelo
summary(modelo)

Si nos guiamos únicamente por los resultados obtenidos en este modelo, el R^2 ajustado resulta de 0,975, lo cual es bastante alto.
Entre los coeficientes, hay 14 de ellos que son significativamente no nulos con un nivel del 95%:
Intercept
CHEST1
KNEE
SHOUL
CHESTG
WAISTG
HIP
GLUTE
FLOREA
KNEEG
CALF
AGE
HEIG
GEN

Respecto al estadístico F, el test resulta significativo a cualquier nivel de significación típico, por lo que se de descarta que todos los coeficientes sean nulos (al menos alguno es significativamente distinto de cero, es decir, se puede utilizar para estimar el peso.

In [ ]:
# Crea un heatmap

datos_entrenamiento_2 <- datos_entrenamiento[, !colnames(datos_entrenamiento) %in% "TrainTest"]

heatmap(cor(datos_entrenamiento_2), #matriz de correlación
        col = colorRampPalette(c("blue", "white", "red"))(20),
        main = "Heatmap de Correlaciones",
        Rowv = NA, Colv = NA)


Hay 3 variables en particular que parecen estar muy correlacionadas con las otras: AGE, GLUTE y BIIL. Pero, a su vez estas 3 variables no parecen correlacionadas entre sí.

FALTA RESPONDER.
¿Que efecto produce la presencia de multicolinealidad en los errores de estimacion de los coeficientes?



In [ ]:
# Crea un heatmap sin AGE

datos_entrenamiento_sin_age <- datos_entrenamiento_2[, !colnames(datos_entrenamiento_2) %in% "AGE"]


heatmap(cor(datos_entrenamiento_sin_age), #matriz de correlación
        col = colorRampPalette(c("blue", "white", "red"))(20),
        main = "Heatmap de Correlaciones",
        Rowv = NA, Colv = NA)

In [ ]:
install.packages("GGally")
library(GGally)

In [ ]:
ggpairs (data = df)

In [ ]:
# Crea un heatmap
heatmap.2(cor(df), #matriz de correlación
        #col="heat.colors",
        key = FALSE,
        col = colorRampPalette(c("blue", "white", "red"))(20),
        main = "Heatmap de Correlaciones",
        Rowv = NULL, Colv = NULL)


In [ ]:
pairs(df)

In [ ]:
# Error empírico del modelo en el testeo

predicciones <- predict(modelo, newdata = datos_prueba)

# Calcular el error de predicción empírico (por ejemplo, el error cuadrático medio)
error_empirico <- mean((datos_prueba$WEIG - predicciones)^2)

# Mostrar el resultado
print(paste("Error de predicción empírico en el grupo de testeo:", error_empirico))



# (h) Nuevo modelo con algunas variables seleccionadas

In [ ]:
modelizar <- function (predictores, respuesta, datos_entrenamiento, datos_prueba) {
    # Crea el modelo lineal
    modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+")), data = datos_entrenamiento)

    # Muestra un resumen del modelo
    print(summary(modelo))

    #A partir de acá se usan datos_prueba:

    # Realiza predicciones en los datos de prueba
    predicciones_prueba <- predict(modelo, newdata = datos_prueba)

    # Calcular el error de predicción empírico (error cuadrático medio)
    error_empirico <- mean((datos_prueba$WEIG - predicciones_prueba)^2)

    # Calcula el R cuadrado
    r_cuadrado <- 1 - sum((datos_prueba$WEIG - predicciones_prueba)^2) / sum((datos_prueba$WEIG - mean(datos_prueba$WEIG))^2)

    # Calcula el número de predictores en el modelo
    num_predictores <- length(predictores)

    # Calcula el número de observaciones en los datos de prueba
    num_obs_prueba <- nrow(datos_prueba)

    # Calcula el R cuadrado ajustado
    r_cuadrado_ajustado <- 1 - ((1 - r_cuadrado) * (num_obs_prueba - 1)) / (num_obs_prueba - num_predictores - 1)

    # Muestra los resultados
    resultados_prueba <- data.frame(Predicted_WEIG = predicciones_prueba, R_cuadrado = r_cuadrado, R_cuadrado_ajustado = r_cuadrado_ajustado)
    #print(resultados_prueba)

    # Mostrar los resultados
    cat("R^2 en el grupo de testeo:", r_cuadrado, '\n')
    cat("R^2 ajustado en el grupo de testeo", r_cuadrado_ajustado, '\n')
    cat("Error de predicción empírico en el grupo de testeo:", error_empirico)
}

In [ ]:
#las significativas del anterior

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'HIP', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'AGE', 'HEIG', 'GEN')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

el R2 ajustado para el grupo de entrenamiento es igual que en el modelo anterior con todas las variables: 0.975

Error de predicción empírico en el grupo de testeo también es muy similar: 4.22 con todas las variables y 4.20 con las variables significativas

In [ ]:
# Crea un heatmap de estas variables seleccionadas

df_predictores <- df[, predictores]


heatmap(cor(df_predictores), #matriz de correlación
        col = colorRampPalette(c("blue", "white", "red"))(20),
        main = "Heatmap de Correlaciones",
        Rowv = NA, Colv = NA)

AGE y GLUTE parecen muy correlacionadas con las otras. Probamos sin ellas:

In [ ]:
#las significativas del anterior, sin GLUTE

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'HIP', 'FLOREA', 'KNEEG', 'CALF', 'AGE', 'HEIG', 'GEN')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'HIP', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'HEIG', 'GEN')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE ni GLUTE

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'HIP', 'FLOREA', 'KNEEG', 'CALF', 'HEIG', 'GEN')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

Sin AGE ni GLUTE el R2 ajustado en el grupo de testeo se mantiene similar: 0.97

In [ ]:
# Crea un heatmap de estas variables seleccionadas

df_predictores <- df[, predictores]


heatmap(cor(df_predictores), #matriz de correlación
        col = colorRampPalette(c("blue", "white", "red"))(20),
        main = "Heatmap de Correlaciones",
        Rowv = NA, Colv = NA)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'HIP', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
# Crea un heatmap de estas variables seleccionadas

df_predictores <- df[, predictores]


heatmap(cor(df_predictores), #matriz de correlación
        col = colorRampPalette(c("blue", "white", "red"))(20),
        main = "Heatmap de Correlaciones",
        Rowv = NA, Colv = NA)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin GLUTE

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'FLOREA', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin HEIG

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1

predictores <- c('KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin KNEE

predictores <- c('SHOUL', 'CHESTG', 'WAISTG', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin SHOUL

predictores <- c('KNEE', 'CHESTG', 'WAISTG', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin SHOUL, sin CHESTG

predictores <- c('KNEE', 'WAISTG', 'GLUTE', 'FLOREA', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin SHOUL, sin FLOREA

predictores <- c('KNEE', 'CHESTG', 'WAISTG', 'GLUTE', 'KNEEG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#MODO GOD ALE

#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin SHOUL, sin FLOREA, sin KNEEG

predictores <- c('KNEE', 'CHESTG', 'WAISTG', 'GLUTE', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
# Crea un heatmap de estas variables seleccionadas

df_predictores <- df[, predictores]


heatmap(cor(df_predictores), #matriz de correlación
        col = colorRampPalette(c("blue", "white", "red"))(20),
        main = "Heatmap de Correlaciones",
        Rowv = NA, Colv = NA)

In [ ]:
#Al sacar GLUTE empeora....

#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin SHOUL, sin FLOREA, sin KNEEG

predictores <- c('KNEE', 'CHESTG', 'WAISTG', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#MODO GOD ALE

#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin SHOUL, sin FLOREA, sin KNEEG

predictores <- c('KNEE', 'CHESTG', 'WAISTG', 'GLUTE', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+")), data = datos_entrenamiento)


plot(modelo)

In [ ]:


#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CHEST1, sin SHOUL, sin FLOREA, sin KNEEG

predictores <- c('KNEE', 'CHESTG', 'WAISTG', 'GLUTE', 'CALF', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
#las significativas del anterior, sin AGE, sin GEN, sin HIP, sin CALF

predictores <- c('CHEST1', 'KNEE', 'SHOUL', 'CHESTG', 'WAISTG', 'FLOREA', 'KNEEG', 'HEIG')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)

In [ ]:
# Mod God Echi
predictores <- c('SHOUL', 'CHESTG', 'WAISTG', 'HIP','CALF','KNEE','ELBOW','CHEST1','AGE')
respuesta <- 'WEIG'

modelizar (predictores, respuesta, datos_entrenamiento, datos_prueba)


### Mejor Modelo porque tiene muy buen r con pocas variables

In [ ]:
# Selecciona las variables predictoras y la variable de respuesta
predictores <- c('WAISTG','AGE')
respuesta <- 'WEIG'

# Crea el modelo lineal
modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+"),-1), data = df)

# Muestra un resumen del modelo
summary(modelo)

### Mejor modelo por r

In [ ]:
# Selecciona las variables predictoras y la variable de respuesta
predictores <- c('WAISTG','AGE','GEN','GLUTE')
respuesta <- 'WEIG'

# Crea el modelo lineal
modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+"),-1), data = datos)

# Muestra un resumen del modelo
summary(modelo)

### Modelo con 5 variables

In [ ]:
# Selecciona las variables predictoras y la variable de respuesta
predictores <- c('CHEST2', 'SHOUL', 'CHESTG', 'WAISTG', 'GLUTE','GEN')
respuesta <- 'WEIG'

# Crea el modelo lineal
modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+")), data = datos)

# Muestra un resumen del modelo
summary(modelo)

# 6 variables

In [ ]:
# Selecciona las variables predictoras y la variable de respuesta
predictores <- c('CHEST2', 'SHOUL', 'CHESTG', 'WAISTG', 'NAVEL', 'GLUTE')
respuesta <- 'WEIG'

# Crea el modelo lineal
modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+")), data = datos)

# Muestra un resumen del modelo
summary(modelo)

# 7 variables

In [ ]:
# Selecciona las variables predictoras y la variable de respuesta
predictores <- c('CHEST2', 'SHOUL', 'CHESTG', 'WAISTG', 'GLUTE','GEN','WRISTG')
respuesta <- 'WEIG'

# Crea el modelo lineal
modelo <- lm(formula = paste(respuesta, "~", paste(predictores, collapse = "+")), data = datos)

# Muestra un resumen del modelo
summary(modelo)

#(i) Glmnet

In [ ]:
#Todas las variables
predictores <- c("BIAC","BIIL","BITRO","CHEST1","CHEST2","ELBOW","WRIST", "KNEE","ANKLE","SHOUL","CHESTG","WAISTG","NAVEL","HIP","GLUTE","BICEP", "FLOREA","KNEEG","CALF","ANKLEG","WRISTG","AGE","HEIG","GEN")
respuesta <- 'WEIG'


# Ajusta un modelo LASSO con glmnet
modelo_lasso <- glmnet(as.matrix(datos_entrenamiento[, predictores]), datos_entrenamiento$WEIG)

# Imprime la tabla de coeficientes para distintos valores de lambda
print(coef(modelo_lasso))

# Grafica los coeficientes en función de lambda
plot(modelo_lasso, xvar = "lambda", label = TRUE)

Falta interpretar lo anterior
¿cuál es el estimador generalizado nuevo?

# (j) Lambda óptimo

In [ ]:
tabla_coeficientes <- coef(modelo_lasso)
#print(tabla_coeficientes)

# Calcula y almacena el R cuadrado para cada valor de lambda en los datos de prueba
r_cuadrado_por_lambda <- numeric(length(modelo_lasso$lambda))
for (i in seq_along(modelo_lasso$lambda)) {
  predicciones_prueba <- predict(modelo_lasso, newx = as.matrix(datos_prueba[, predictores]), s = modelo_lasso$lambda[i])
  r_cuadrado_por_lambda[i] <- 1 - sum((datos_prueba$WEIG - predicciones_prueba)^2) / sum((datos_prueba$WEIG - mean(datos_prueba$WEIG))^2)
}


resultados_lasso <- data.frame(lambda = modelo_lasso$lambda, R_cuadrado = r_cuadrado_por_lambda)

par(mfrow = c(1, 2))
plot(modelo_lasso, xvar = "lambda")
plot(resultados_lasso$lambda, resultados_lasso$R_cuadrado, type = "l", xlab = "Lambda", ylab = "R cuadrado", main = "R cuadrado vs Lambda")

In [ ]:
cv_modelo <- cv.glmnet(as.matrix(datos_entrenamiento[, predictores]), datos_entrenamiento$WEIG, alpha = 1)

# Muestra el valor óptimo de lambda según el criterio de 1 desvío estándar
lambda_optimo <- cv_modelo$lambda.min
print(paste("Valor óptimo de lambda:", lambda_optimo))

# Ajusta el modelo con el valor óptimo de lambda
modelo_lasso_optimo <- glmnet(as.matrix(datos_entrenamiento[, predictores]), datos_entrenamiento$WEIG, alpha = 1, lambda = lambda_optimo)

# Imprime la tabla de coeficientes
print(coef(modelo_lasso_optimo))

# Calcula y almacena el R cuadrado en los datos de prueba
predicciones_prueba_optimo <- predict(modelo_lasso_optimo, newx = as.matrix(datos_prueba[, predictores]))
r_cuadrado_optimo <- 1 - sum((datos_prueba$WEIG - predicciones_prueba_optimo)^2) / sum((datos_prueba$WEIG - mean(datos_prueba$WEIG))^2)

# Calcula el error de clasificación empírico en el grupo de testeo
clases_predichas <- predict(modelo_lasso_optimo, newx = as.matrix(datos_prueba[, predictores]), s = lambda_optimo, type = "class")
error_clasificacion_optimo <- mean(clases_predichas != datos_prueba$WEIG)
print(paste("Error de clasificación empírico en el grupo de testeo:", error_clasificacion_optimo))